### PR TITLE
Add AcceptOffers Support

### PR DESCRIPTION
**Pull request by drewkerrigan**
Addressing #187 

Added `AcceptOffers` to the scheduler driver which utilizes the new HTTP API to send a Scheduler Call Message to the Mesos Master. Supported operation types are `LAUNCH`, `RESERVE`, `UNRESERVE`, `CREATE`, `DESTROY`.

I understand that work is currently happening in the [next](https://github.com/mesos/mesos-go/tree/next) branch, but until that is done, this is a way to utilize dynamic reservations / persistent volumes in the existing mesos-go binding.

### CODE REVIEW

File: messenger/messenger.go [hunk 1]
```diff
@@ -391,5 +391,15 @@ func (m *MesosMessenger) decodeLoop() {
 
 // getMessageName returns the name of the message in the mesos manner.
 func getMessageName(msg proto.Message) string {
-	return fmt.Sprintf("%v.%v", "mesos.internal", reflect.TypeOf(msg).Elem().Name())
+	msgType := reflect.TypeOf(msg).Elem().Name()
+	var msgName string
+
+	switch msgType {
```
Comment (jdef) [inline messenger/messenger.go]: i'd rather see a type-switch vs. reflection:
```go
switch msg := msg.(type); msg {
case *mesos.Call:
  // ... msg is now a *mesos.Call instance
default:
  // msg could be anything else
}
```

File: messenger/message.go [hunk 2]
```diff
@@ -36,7 +36,29 @@ type Message struct {
...
+	case "external":
+		requestURI = fmt.Sprintf("/api/v1/%s", m.Name)
+	default:
+		requestURI = fmt.Sprintf("/%s/%s", m.UPID.ID, m.Name)
+	}
+
+	return requestURI
+}
+
+func (m *Message) GetType() string {
```
Comment (jdef) [inline messenger/message.go]: would rather see:
```go
func (m *Message) isV1API() bool {
...
}
```

File: scheduler/scheduler.go [hunk 3]
```diff
@@ -1091,6 +1091,110 @@ func (driver *MesosSchedulerDriver) abort(cause error) (stat mesos.Status, err e
...
+
+	okOperations := make([]*mesos.Offer_Operation, 0, len(operations))
+
+	for _, offerId := range offerIds {
+		for _, operation := range operations {
+			// Keep only the slave PIDs where we run tasks so we can send
+			// framework messages directly.
+			if driver.cache.containsOffer(offerId) {
+				// Validate
+				switch *operation.Type {
```
Comment (jdef) [inline scheduler/scheduler.go]: s/*operation.Type/operation.GetType()/

File: scheduler/scheduler.go [hunk 4]
```diff
@@ -1091,6 +1091,110 @@ func (driver *MesosSchedulerDriver) abort(cause error) (stat mesos.Status, err e
...
+			} else {
+				log.Warningf("Attempting to accept offers with unknown offer %s\n", offerId.GetValue())
+			}
+		}
+
+		driver.cache.removeOffer(offerId) // if offer
+	}
+
+	// Accept Offers
+	accept := &mesos.Call_Accept{
```
Comment (neilconway) [inline scheduler/scheduler.go]: I think it is a bit better to construct this in a single literal:

```
message := &mesos.Call{
    FrameworkId: ...,
    Type: mesos.Call_ACCEPT.Enum(),
    Accept: &mesos.Call_Accept{
        ...
    },
}```

File: scheduler/scheduler.go [hunk 5]
```diff
@@ -1091,6 +1091,110 @@ func (driver *MesosSchedulerDriver) abort(cause error) (stat mesos.Status, err e
...
+		return driver.status, fmt.Errorf("Not connected to master.")
+	}
+
+	okOperations := make([]*mesos.Offer_Operation, 0, len(operations))
+
+	for _, offerId := range offerIds {
+		for _, operation := range operations {
+			// Keep only the slave PIDs where we run tasks so we can send
+			// framework messages directly.
+			if driver.cache.containsOffer(offerId) {
```
Comment (neilconway) [inline scheduler/scheduler.go]: I think it is a bit clearer to invert the conditional:

```
if !driver.cache.containsOffer(offerId) {
    // warning
    continue
}

// Validate
```

**Comment by drewkerrigan** [comment]
@neilconway @jdef Good feedback, I've addressed your comments in the latest commit.

File: scheduler/scheduler.go [hunk 6]
```diff
@@ -1091,6 +1091,109 @@ func (driver *MesosSchedulerDriver) abort(cause error) (stat mesos.Status, err e
...
+
+		driver.cache.removeOffer(offerId) // if offer
+	}
+
+	// Accept Offers
+	message := &mesos.Call{
+		FrameworkId: driver.frameworkInfo.Id,
+		Type:        mesos.Call_ACCEPT.Enum(),
+		Accept: &mesos.Call_Accept{
+			OfferIds:   offerIds,
```
Comment (jdef) [inline scheduler/scheduler.go]: It's possible that `offerIds` will contain ID's that are not actually used (because of the earlier `continue`)

Comment (jdef) [inline scheduler/scheduler.go]: nevermind. misread code

File: scheduler/scheduler.go [hunk 7]
```diff
@@ -1091,6 +1091,109 @@ func (driver *MesosSchedulerDriver) abort(cause error) (stat mesos.Status, err e
...
 
+func (driver *MesosSchedulerDriver) AcceptOffers(offerIds []*mesos.OfferID, operations []*mesos.Offer_Operation, filters *mesos.Filters) (mesos.Status, error) {
+	driver.eventLock.Lock()
+	defer driver.eventLock.Unlock()
+
+	if stat := driver.status; stat != mesos.Status_DRIVER_RUNNING {
+		return stat, fmt.Errorf("Unable to AcceptOffers, expected driver status %s, but got %s", mesos.Status_DRIVER_RUNNING, stat)
+	}
+
+	if !driver.connected {
```
Comment (jdef) [inline scheduler/scheduler.go]: actually you probably want to send TASK_LOST to the Scheduler for each LAUNCH op. see launchTasks() and `pushLostTask`. the native code does this: https://github.com/apache/mesos/blob/master/src/sched/sched.cpp#L1160

Comment (drewkerrigan) [inline scheduler/scheduler.go]: Ok, this is added

**Comment by drewkerrigan** [comment]
Ah yeah, good catch, adding

**Comment by drewkerrigan** [comment]
ping @jdef @neilconway, I added some unit tests as well as an example scheduler that uses the new functionality. Let me know if you need additional changes, or if anything else is preventing a merge.

**Comment by jdef** [comment]
lgtm :ship:

**Pull request merged**